### ISSUE TITLE
Docrepr doesn't understand Sphinx ":type" and ":param" directives?

### ISSUE THREAD

**Issue by fasiha** [spyder-ide/docrepr#34]
I was following the Sphinx tutorial's section on [Automatic documentation generation from code](https://www.sphinx-doc.org/en/master/tutorial/automatic-doc-generation.html), which has the following docstring:
```py
def get_random_ingredients(kind=None):
    """
    Return a list of random ingredients as strings.

    :param kind: Optional "kind" of ingredients. <----------- !!
    :type kind: list[str] or None  <------------------------- !!
    :raise lumache.InvalidKindError: If the kind is invalid.
    :return: The ingredients list.
    :rtype: list[str]

    """
    return ["eggs", "bacon", "spam"]
```
Notice how it uses `:param:` and `:type` to document both the name and the type of a parameter. When Sphinx renders this docstring (via the `autodoc` extension), it looks great:

<img width="590" alt="Screen Shot 2022-01-11 at 10 46 21 PM" src="https://user-images.githubusercontent.com/37649/149077743-f1b409a3-022d-4d38-8d30-d1307860b733.png">

The above is all about Sphinx.

Now to Docrepr.

When I attempt to render the same docstring with Docrepr, those two directives don't seem to be parsed properly. For example, if I
1. `pip install /path/to/master/branch/docrepr`
2. `git clone https://github.com/fasiha/lumache.git && cd lumache`
3. `python test_docrepr.py`

where `test_docrepr.py` contains the following:
```py
from docrepr import sphinxify                 # html generator
from IPython.core.oinspect import Inspector   # oinfo generator

import lumache

oinfo = Inspector().info(lumache.get_random_ingredients)
oinfo['name'] = 'get_random_ingredients'
url = sphinxify.rich_repr(oinfo)
print('file://' + url)
```
the output file looks like this:

<img width="368" alt="Screen Shot 2022-01-11 at 10 50 27 PM" src="https://user-images.githubusercontent.com/37649/149078177-4b794e04-8fa0-423d-817b-440e94464267.png">

It looks like the two directives, `param` and `type` both got their own bullet, as if Sphinx didn't recognize that they describe the same parameter?

Is there something special that needs to go into `conf.py` for Sphinx to understand this directive? I tried adding a line to conf.py,
```py
extensions = [
    'sphinx.ext.autodoc',
    'sphinx.ext.autosummary',
]
```
but this didn't change the output of Docrepr.

Am I right in my guess that Docrepr's conf.py needs to be tweaked for this to work? Are there other Sphinx/reST directives that Docrepr doesn't understand?

Thanks for taking a look!

**Comment by martinRenou** [comment]
I've looked into this but I can't really understand why sphinx is not generating the docs properly. 

Even when generating a text output it will not behave:
```
Return a list of random ingredients as strings.

param kind:
   Optional "kind" of ingredients.

type kind:
   list[str] or None

raise lumache.InvalidKindError:
   If the kind is invalid.

return:
   The ingredients list.

rtype:
   list[str]
```

I can also reproduce the fact that, even when autodoc/doctest extensions are added, it does not do what the Sphinx docs advertises.

**Comment by CAM-Gerlach** [comment]
Note that the problem appears to affect all the Sphinx docstring directives, not just param, type, etc.

Just for completeness, have you tried Googledoc and/or Numpydoc format docstrings? Spyder renders Sphinx and Googledoc style docstrings the same as docrepr does, but Numpydoc docstrings render as desired. I can't remember exactly what custom handling we have for that case, but we can look into it.

> I can also reproduce the fact that, even when autodoc/doctest extensions are added, it does not do what the Sphinx docs advertises.

Your knowledge runs deeper than mine, but at a high level, what I suspect could be the reason for the difference is the fact that when running Sphinx on a codebase with the sphinx-autodoc extension, it recognizes the files as Python code which triggers the appropriate autodoc parsing of the embedded docstrings (in fact, autodoc _imports_ the modules to be documented), whereas with docrepr, we're just feeding Sphinx the raw docstring as a pseduo-rst file, so autodoc doesn't know it needs to parse it. 

Autodoc relies, AFAIK, at least to some extent on the structure and content in the original source code to discover and generate its own document structure, links and I believe at least some postprocessing of names, etc. (and imports it, after all), so assuming that's what is going on, I'm not sure if there's a more efficient approach than running having docrepr run sphinx + autodoc on the entire codebase and then extracting the resulting docs for the name of interest (which wouldn't work when run against dynamically rather than statically).

There may be a way to tell Sphinx/autodoc to interpret a passed reST as intermediate extracted autodoc output, or to trigger autodoc's preprocessing, perhaps by calling its Python API/functions directly, or even by some `conf.py` options, but looking around online I couldn't really find much about this. We can also look into how Spyder currently handles it, though IIRC its custom stuff for Numpydoc that hasn't been touched in a long time (and it would benefit Spyder to support Googledoc and Sphinxdoc too, just like docrepr, since they both stem from a common root).

**Comment by fasiha** [comment]
> when running Sphinx on a codebase with the sphinx-autodoc extension, it recognizes the files as Python code which triggers the appropriate autodoc parsing of the embedded docstrings (in fact, autodoc _imports_ the modules to be documented), whereas with docrepr, we're just feeding Sphinx the raw docstring as a pseduo-rst file, so autodoc doesn't know it needs to parse it

Thanks for the insight @CAM-Gerlach! I wanted to comment on this because, Sphinx *is* aware of the `param` and `type` outside a docstring in pure reST: the previous step of the tutorial, [Describing code in Sphinx](https://www.sphinx-doc.org/en/master/tutorial/describing-code.html) shows the `py:function` directive using `type` and `param`:
```rest
Creating recipes
----------------

To retrieve a list of random ingredients,
you can use the ``lumache.get_random_ingredients()`` function:

.. py:function:: lumache.get_random_ingredients(kind=None)

   Return a list of random ingredients as strings.

   :param kind: Optional "kind" of ingredients.
   :type kind: list[str] or None
   :return: The ingredients list.
   :rtype: list[str]
```
which Sphinx renders without any further Python-specific information:

![](https://www.sphinx-doc.org/en/master/_images/lumache-py-function.png)

I wonder if we can wrap the raw docstring in the pseudo-reST file in `py:function`?

> have you tried Googledoc and/or Numpydoc format docstrings

I have not, I've mainly been testing with docstrings that Sphinx handles well on its own

> even when autodoc/doctest extensions are added

Quick note @martinRenou I forgot that Docrepr has a `generate_extensions` that overrides the extensions in `conf.py` but even if I add `autodoc` here it doesn't help. https://github.com/spyder-ide/docrepr/blob/ce67c9a4b8535f9732b640040bec5e1ecabb0a32/docrepr/sphinxify.py#L200-L204

**Comment by CAM-Gerlach** [comment]
Ah, thanks for the detailed followup, @fasiha !

It seems what's really going on is that `:param:`, `:type:`, `:return:` etc. are actually options to the `py:function::` directive, rather than roles on their own, and Sphinx thus naturally simply doesn't know what to do with them outside of that context. When Sphinx-Autodoc runs on a codebase, what its presumably doing under the hook is something like extracting the docstrings it finds and placing them inside `:py:<type>` blocks with the function/class signature in an intermediate `.rst` file, which Sphinx then renders (how Napoleon and other similar tools generally work, as I understand, is by converting the googledoc, numpydoc, etc to the equivalent `:py:<type>` representation after extraction but before Sphinx rendering).

So, we don't need autodoc to solve this at all; rather, this should be possible by wrapping the docstring in an appropriate `:py:<type>` directive ourselves.

There are some potentially tricky edge cases, but I've hacked out a basic proof of concept implementation that should work for the great majority of Sphinx-format docstrings, while leaving others unaffected. See the [`add-sphinxdoc-rendering` branch on my fork](https://github.com/CAM-Gerlach/docrepr/tree/add-sphinxdoc-rendering). Here's what it looks like for your example function:

![image](https://user-images.githubusercontent.com/17051931/149434040-672e7e16-9e91-4d0a-911a-9bb57c6335b9.png)

You can test it yourself by cloning my fork and running in the root:

```bash
pytest -k function --open-browser
```

I can go ahead with a preliminary PR, if you all like the approach.

**Comment by fasiha** [comment]
Aha! Superb!!! Well done @CAM-Gerlach, looks great to me! I will check out your branch soon as possible and play with it but right now, the only change I notice is it now includes the signature of the function at the top—should be fine.

I'm excited to check out your branch to make sure module/class docstrings still look good too (e.g., if you print the info for `numpy.sin`).

@martinRenou what do you think?

**Comment by CAM-Gerlach** [comment]
> right now, the only change I notice is it now includes the signature of the function at the top—should be fine.

Yeah, I noticed that; on one hand its useful since in at least some cases the signature is not printed at the top, but we'll want to tweak the CSS to make it look better if we include it, or we can just hide that line if its duplicative.

> I'm excited to check out your branch to make sure module/class docstrings still look good too (e.g., if you print the info for numpy.sin).

I tested `numpy.sin` (and the rest of the examples in the test suite), and it is not harmed; the new code applies a number of heuristics to ensure only docstring with `:py:` options/markup are formatted as such, while leaving others as they are. Numpydoc docstrings like `numpy.sin` are already rendered much more nicely due to custom Numpydoc rendering; it would be cool to use that for the Sphinx docstrings too, but that's more of a stretch goal. 

I made some additional fixes and cleanup, particularly surrounding use with non-function objects, in case you've been testing it already.
